horovod/common/operations.cc

```diff
@@ -995,68 +1002,162 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
+        RECORD_EVENT(entries, event_queue, MEMCPY_IN_HOST_BUFFER, stream)
 
-          ACTIVITY_START_ALL(entries, timeline, MPI_ALLREDUCE)
-          MPI_CHECK(entries, "MPI_Allreduce",
-                    MPI_Allreduce(MPI_IN_PLACE, host_buffer, (int)num_elements,
-                                  GetMPIDataType(first_entry.tensor), MPI_SUM,
-                                  horovod_global.cross_comm))
-          ACTIVITY_END_ALL(entries, timeline)
+        // Synchronize.
+        WAIT_FOR_EVENTS(entries, timeline, event_queue)
 
-          CUDA_CHECK(entries, "cudaMemcpyAsync",
-                     cudaMemcpyAsync(buffer_data, host_buffer, buffer_len,
-                                     cudaMemcpyHostToDevice, stream))
-          if (timeline.Initialized()) {
-            RECORD_EVENT(entries, event_queue, MEMCPY_OUT_HOST_BUFFER, stream)
-          }
+        ACTIVITY_START_ALL(entries, timeline, MPI_ALLREDUCE)
+        MPI_CHECK(entries, "MPI_Allreduce",
+                  MPI_Allreduce(MPI_IN_PLACE, host_buffer,
+                                (int)num_elements_per_rank,
+                                GetMPIDataType(first_entry.tensor), MPI_SUM,
+                                horovod_global.cross_comm))
+
+        ACTIVITY_END_ALL(entries, timeline)
+
+        CUDA_CHECK(entries, "cudaMemcpyAsync",
+                   cudaMemcpyAsync(buffer_data_at_rank_offset, host_buffer,
+                                   buffer_len_per_rank, cudaMemcpyHostToDevice,
+                                   stream))
+        if (timeline.Initialized()) {
+          RECORD_EVENT(entries, event_queue, MEMCPY_OUT_HOST_BUFFER, stream)
         }
 
-        NCCL_CHECK(entries, "ncclBcast",
-                   ncclBcast(buffer_data, (size_t)num_elements,
-                             GetNCCLDataType(first_entry.tensor), 0, nccl_comm,
-                             stream))
+        NCCL_CHECK(entries, "ncclAllGather",
+                   ncclAllGather(buffer_data_at_rank_offset, buffer_data,
+                                 (size_t)num_elements_per_rank,
+                                 GetNCCLDataType(first_entry.tensor), nccl_comm,
+                                 stream))
+
         if (timeline.Initialized()) {
-          RECORD_EVENT(entries, event_queue, NCCL_BCAST, stream)
+          RECORD_EVENT(entries, event_queue, NCCL_ALLGATHER, stream)
+        }
+
+        // Synchronize.
+        WAIT_FOR_EVENTS(entries, timeline, event_queue)
+
+        if (num_elements_remaining > 0) {
+
+          // For the remaining data do ncclReduce - MPIAllreduce - ncclBcast
+          void* buffer_data_remainder =
+              (uint8_t*)buffer_data +
+              buffer_len_per_rank * horovod_global.local_size;
+          void* fused_input_data_remainder =
+              (uint8_t*)fused_input_data +
+              buffer_len_per_rank * horovod_global.local_size;
+
+          NCCL_CHECK(entries, "ncclReduce",
```

Comment:
I've tested that memcpy is not needed without host_buffer

Reply:
@anpark, please see comment above related to host_buffer.

Reply:
@alsrgv could you make it clear? I don't found that comment.

Reply:
@alsrgv The performance is slightly worse than the earlier version (by about 4-5%), seemingly due to the disproportionate amount of time taken by the tiny allreduce on the remaining data (see attached figure). Still much faster than the existing version though.

Pipelining actually makes a lot of sense not only to reduce synchronization but also because we can combine the two MPI calls into a single one (just a slightly larger one for `local_rank==0`), thereby removing some of the overhead. I can try this next.

My only concern is that it seemed to me like back-to-back NCCL calls are not registered in timeline properly without synchronization in between, do you have any idea why this could be the case?

<img width="982" alt="screen shot 2018-08-09 at 10 38 08 pm" src="https://user-images.githubusercontent.com/11875690/43940648-73ed4c60-9c25-11e8-884b-fc54c5a69cc2.png">

Reply:
not clear about "back-to-back NCCL calls are not registered in timeline properly without synchronization in between"

Reply:
@alsrgv the deadlock problem happened when concurrency, but we use WAIT_FOR_EVENT to sync.

Reply:
@karakusc, I believe if you do back-2-back calls but still register event between them, you should be able to separate them in the timeline.  Is that not what you're experiencing?

@anpark, we've seen deadlock happen in tickets referenced in #435.

Reply:
Made this change. Performance is slightly better now, the only synchronization is right before the MPI Allreduce call, and the timeline shows each NCCL call separately (whenever the data is non-divisible).

Reply:
One outstanding issue is that this algorithm will not work for heterogeneous clusters (different `local_size` across nodes). Maybe we can try reducing the local tensors at a set of representatives who do the cross-node MPI Allreduce, and the number of representatives = min. `local_size` across nodes. Alternatively we can simply fall back to existing hierarchical allreduce in the heterogeneous case to keep things simple.